src/sentry/tasks/relocation.py

```diff
@@ -456,18 +492,143 @@ def preprocessing_complete(uuid: str) -> None:
+    try:
+        return RelocationValidationAttempt.objects.get(
+            relocation=relocation, relocation_validation=relocation_validation, build_id=build_id
+        )
+    except RelocationValidationAttempt.DoesNotExist:
+        fail_relocation(
+            relocation,
+            task,
+            ERR_VALIDATING_ATTEMPT_MISSING,
+        )
+        return None
+
+
+def _update_relocation_validation_attempt(
+    task: OrderedTask,
+    relocation: Relocation,
+    relocation_validation: RelocationValidation,
+    relocation_validation_attempt: RelocationValidationAttempt,
+    status: ValidationStatus,
+) -> None:
+    """
+    After a `RelocationValidationAttempt` resolves, make sure to update the owning
+    `RelocationValidation` and `Relocation` as well.
+    """
+
+    with atomic_transaction(
+        using=(
+            router.db_for_write(Relocation),
+            router.db_for_write(RelocationValidation),
+            router.db_for_write(RelocationValidationAttempt),
+        )
+    ):
+        # If no interesting status updates occurred, check again in a minute.
+        if status == ValidationStatus.IN_PROGRESS:
+            logger.info(
+                "Validation polling: scheduled",
+                extra={"uuid": relocation.uuid, "task": task.name},
+            )
+            validating_poll.apply_async(
+                args=[relocation.uuid, str(relocation_validation_attempt.build_id)], countdown=60
+            )
+            return
+
+        relocation_validation_attempt.status = status.value
+
+        # These statuses merit failing this attempt and kicking off a new
+        # `RelocationValidationAttempt`, if possible.
+        if status in {ValidationStatus.TIMEOUT, ValidationStatus.FAILURE}:
+            if relocation_validation.attempts < MAX_VALIDATION_POLL_ATTEMPTS:
+                relocation_validation_attempt.status = status.value
+                relocation_validation_attempt.save()
+
+                relocation.latest_task = OrderedTask.VALIDATING_START.name
+                relocation.save()
+
+                logger.info(
+                    "Validation timed out",
+                    extra={"uuid": relocation.uuid, "task": task.name},
+                )
+                validating_start.delay(relocation.uuid)
+                return
+
+            # Always accept the numerically higher `ValidationStatus`, since that is a more definite
+            # result.
+            if relocation_validation.status < status.value:
+                relocation_validation.status = status.value
+                relocation_validation_attempt.save()
+            return fail_relocation(
+                relocation, task, "Validation could not be completed. Please contact support."
+            )
+
+        # All remaining statuses are final, so we can update the owning `RelocationValidation` now.
+        assert status in {ValidationStatus.INVALID, ValidationStatus.VALID}
```

Comment:
Would it be better to throw a more specific error if the status is unexpected here? 

`throw new Error('Expected validation status of INVALID or VALID, received ${status}`

Reply:
I think for this case, where I am trying to state an invariant, `assert` is the more idiomatic way to go? Throwing an exception implies that there is a valid program state that could cause this to be untrue, which is not the case here. The main reason for the `assert` is: 1. Clarity for readers ("you may assume for the rest of this method that that only these statuses are still possible, and 2. To create an invariant that guards against statuses being added/removed.

Reply:
I think you're right, seeing something similar in other areas of code too